src/meta/src/barrier/mod.rs

```diff
@@ -195,57 +245,125 @@ where
+            state.in_flight_prev_epoch = new_epoch;
+            state
+                .update_inflight_prev_epoch(self.env.meta_store())
+                .await
+                .unwrap();
         }
-
         let mut min_interval = tokio::time::interval(self.interval);
         min_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+        let mut barrier_timer: Option<HistogramTimer> = None;
         loop {
             tokio::select! {
                 biased;
                 // Shutdown
                 _ = &mut shutdown_rx => {
-                    tracing::info!("Barrier manager is shutting down");
+                    tracing::info!("Barrier manager inject is shutting down");
                     return;
                 }
+                result = barrier_result_rx.recv() =>{
+                    match result{
+                        Some(BarrierCommitResult::Ok{new_epoch,actors_to_finish,notifiers,responses}) => {
+                            tracker.add(new_epoch, actors_to_finish, notifiers);
+                            for progress in responses.into_iter().flat_map(|r| r.create_mview_progress) {
+                                tracker.update(progress);
+                            }
+                        }
+                        Some(BarrierCommitResult::Err{new_epoch,err_msg}) => {
+                            if self.enable_recovery {
+                                // If failed, enter recovery mode.
+                                let (new_epoch, actors_to_track, create_mview_progress) =
+                                self.recovery(new_epoch).await;
+                                tracker = CreateMviewProgressTracker::default();
+                                tracker.add(new_epoch, actors_to_track, vec![]);
+                                for progress in create_mview_progress {
+                                    tracker.update(progress);
+                                }
+                                state.in_flight_prev_epoch = new_epoch;
+                                state.update_inflight_prev_epoch(self.env.meta_store()).await.unwrap();
+                                self.is_recovery.store(false,atomic::Ordering::SeqCst);
+                            } else {
+                                panic!("failed to execute barrier: {:?}", err_msg);
+                            }
+                        }
+                        _=>{}
+                    }
+                    continue;
+                }
                 // there's barrier scheduled.
-                _ = self.scheduled_barriers.wait_one() => {}
+                _ = self.scheduled_barriers.wait_one() ,if self.can_inject_barrier().await => {
+
+                }
                 // Wait for the minimal interval,
-                _ = min_interval.tick() => {},
+                _ = min_interval.tick() ,if self.can_inject_barrier().await => {
+
+                }
```

Comment:
https://docs.rs/tokio/0.2.18/tokio/macro.select.html

The condition after "if" is a precondition evaluted before checking the async expression "in all branches". Therefore, the current implementation will result in a busy-loop checking on `pause_inject_barrier` even if we didn't reach `self.scheduled_barriers.wait_one()` and `min_interval.tick()`. 

This issue can also be resolved by maintaining a [state machine](https://github.com/singularity-data/risingwave/pull/2184#discussion_r901228497). We can transistion the state machine' state to `PauseInjection` when we reach the in-flight barrier limit and transistion back to normal running state on barrier completion.

Reply:
I do some test about "if". I find there are not "busy-loop checking ".
 I think state machine is a good idea. But I find a questions. The "recovery","build-actor" and "In-flight-nums" will be pause inject simultaneously. So We also use `is_recovery` and so on if we use state machine. I think it is complex. But I don not have good idea to achieve it.